### ISSUE TITLE
Are there jsdoc supported in salsa to polyfill ?

### ISSUE THREAD

**Issue by Thaina** [Microsoft/TypeScript#13167]
Suppose I was implement

```js
Array.prototype.last = function() {
    return this[this.length - 1];
};
```

And wish it present in the intellisense. How should I do?

If it not exist, should we have this feature?

**Comment by aluanhaddad** [comment]
Why not use a `.d.ts` file?
```ts
interface Array<T> {
  last(callbackfn?: (value: T) => boolean): T | undefined;
}
```

**Comment by Thaina** [comment]
First, I need it to be local, only in a file that implement this (and a file and require this file)
Because it written in nodejs. We can't have global code on startup (or can we? I might be wrong here). So there will be a chance that array prototype was not set my polyfill function yet

Second is, I don't like project that contains both typescript and javascript. So I wish to avoid writing my own `.d.ts` that not managed from typings manager. I would like to find salsa way first

**Comment by aluanhaddad** [comment]
You can have  global code on startup. Just import the side effecting code before anything else.
```js
// src/index.js

require('./last').imbue();

// now you have Array.prototype.last available everywhere.
```
```js
// last.js

module.exports = {
  imbue: function () {
    if (Array.prototype.last) return;
    Array.prototype.last = function (predicate) { 
      const source = predicate ? this.filter(predicate) : this;
      return source[source.length - 1]; 
    };
  }
}; 
```

**Comment by Thaina** [comment]
@aluanhaddad require is not global, it could affect only one file. While we may have many file and call `node indexjs` `node default.js` ot `node $filename`

Well actually I was using that solution too because my server has only one endpoint handle all incoming request at the top level

But that still bugs me that it make dependency in 2 places

**Comment by aluanhaddad** [comment]
I don't see how else it could work. If you use a different entry point you will need to polyfill it there as well. I understand why you don't want to do that but I'm not sure that anything can be done, that is just the nature of JavaScript and TypeScript.

**Comment by Thaina** [comment]
@aluanhaddad Well, if it not had planned I wish it would be feature request

We cannot load global things at startup of nodejs application. So I wish that, we could write jsdoc at the implementation point to only polyfill when it really in that file or got required

Suppose I have 3 files. I implement extension of array in first file so I wish to see it in that file
```js
#1st file
/** @memberOf {Array.prototype} < I think memberOf is expected syntax of jsdoc for this feature */
Array.prototype.last = function() { return this[this.length - 1]; }

[0,1,2].las // <<< Wish to see intellisense here

#2nd file
var firstFile = require("./1stFile.js");

["T","e","s","t"].las // <<< Wish to see intellisense here too if possible

#3rd file

["T","e","s","t"].las // <<< Should not see intellisense here because it not implemented and was not made require to any file implement it
```

And the implementation of this feature. Maybe the same as typescript you suggested?
If there are `@memberOf` presented then it would just inject this code
```ts
interface Array<T> { last(callbackfn?: (value: T) => boolean): T | undefined; }
```
Into salsa

Or it impossible?

**Comment by aluanhaddad** [comment]
It depends on the files which are included in the compilation context and on the nature of the global declaration. Since you are using JavaScript, I do not believe you will be able to express this conditionality at all. I think you are right that `memberof` is the correct JSDoc annotation but I do not think it is currently supported.
Personally, I am of the mind that if you need these kinds of highly expressive type constructs (such as module scoped ambient augmentation of a global), you should be using TypeScript. JSDoc sucks at this. One of the reasons TypeScript was created was to solve this problem. JSDoc is underspecified, has multiple competing interpretations, JetBrains, google closure, etc., and is extremely ugly. It also fails to provide correspondence between types and values.

Also, if you use it with google's tools and also with TypeScript, you will run into trouble because closure interprets JSDoc by trying to enforce a convention that JavaScript is a Java-like language while TypeScript interprets JavaScript (correctly) as JavaScript.

**Comment by Thaina** [comment]
@aluanhaddad Well , sometimes it needed and I have no choices
If I could choose I would just use C#. If it will compiles anyway then C# is better than TypeScript

But I'm stuck with javascript and this is not a right time to convert my codebase into C# or typescript

And I think jsdoc itself do excellent work as it could. It then about implementation of each IDE. Even in C#, difference IDE make some difference. It not so standard and I'm fine with stick with a salsa tools when working with js+jsdoc

And so far I think salsa is very refined solution (that's why I choose to use it mainly). I think if salsa would completely implement jsdoc then `@memberOf` should work as this feature I expected

**Comment by aluanhaddad** [comment]
@Thaina 

TLDR; TypeScript is not going anywhere and will survive as long as JavaScript survives

------------------------------------------------------------------------------------------------------------------
TypeScript will always be available, unless npm and GitHub and every other computer with a recent copy dies off. 
To underscore this point, do you know where I transpile my TypeScript most of the time? My web browser.

I host it there and use VS Code for editing, hosting the same version there. TypeScript is a JavaScript program and as long as I have a JavaScript VM available, I know my TypeScript will be code compilable.

**Node**
```sh
PS C:\Source> node
> var ts = require('C:/Users/Aluan/AppData/Roaming/npm/node_modules/typescript');
undefined
> var js = ts.transpileModule("export default function (x: number, y: number): number { return x * y; }", { compilerOptions: { module: ts.ModuleKind.CommonJS } });
undefined
> js
{ outputText: '"use strict";\r\nfunction default_1(x, y) { return x * y; }\r\nexports.__esModule = true;\r\nexports["default"] = default_1;\r\n',
  diagnostics: [],
  sourceMapText: undefined }
> var fs = require('fs');
undefined
> fs.writeFileSync('./multiply.js', js.outputText);
undefined
>
(To exit, press ^C again or type .exit)
>
PS C:\Source> notepad.exe .\multiply.js
PS C:\Source>
```
![image](https://cloud.githubusercontent.com/assets/5034656/21499984/95ffe3b0-cc08-11e6-85b6-db4e9d9411a5.png)


```sh
PS C:\Source> node
> var multiply = require('./multiply.js');
undefined
> multiply.default(4, 6)
24
>
(To exit, press ^C again or type .exit)
>
PS C:\Source>
```

**Browser**
```js
SystemJS.config({packages:{ 'https://cdnjs.cloudflare.com/ajax/libs/typescript/2.0.10/typescript.js' :{format:'global', exports:['ts']}}})

SystemJS.import("https://cdnjs.cloudflare.com/ajax/libs/typescript/2.0.10/typescript.js")
    .then(function () {
        return ts.transpileModule("export default function (x: number, y: number): number { return x * y; }", {
            compilerOptions: {
                module: ts.ModuleKind.CommonJS
            }
        }).outputText;
    }).then(console.log.bind(console));
Promise {[[PromiseStatus]]: "pending", [[PromiseValue]]: undefined}
"use strict";
function default_1(x, y) { return x * y; }
exports.__esModule = true;
exports["default"] = default_1;
```
